examples/gno.land/r/gov/dao/types.gno

```diff
@@ -0,0 +1,232 @@
+	prb.metadata = md
+	return prb
+}
+
+func (prb *ProposalRequestBuilder) WithCallback(cb func(Metadata) error) *ProposalRequestBuilder {
+	prb.callback = cb
+	return prb
+}
+
+func (prb *ProposalRequestBuilder) Build() ProposalRequest {
+	return ProposalRequest{
+		title:       prb.title,
+		description: prb.description,
+		callback:    prb.callback,
+		metadata:    prb.metadata,
+	}
+}
+
+type ProposalRequest struct {
+	title       string
+	description string
+	callback    func(Metadata) error
+	metadata    Metadata
+}
+
+func (p *ProposalRequest) Title() string {
+	return p.title
+}
+
+func (p *ProposalRequest) Description() string {
+	return p.description
+}
+
+func (p *ProposalRequest) Callback() func(Metadata) error {
+	return p.callback
+}
+
+func (p *ProposalRequest) Metadata() Metadata {
+	return p.metadata
+}
+
+type Proposal struct {
+	author Author
+
+	title       string
+	description string
+
+	executor    *executor
+	allowedDAOs []string
+}
+
+func (p *Proposal) Author() Author {
+	return p.author
+}
+
+func (p *Proposal) Title() string {
+	return p.title
+}
+
+func (p *Proposal) Description() string {
+	return p.description
+}
+
+func (p *Proposal) ExecutorString() string {
+	if p.executor != nil {
+		return p.executor.String()
+	}
+
+	return ""
+}
+
+func (p *Proposal) AllowedDAOs() []string {
+	return p.allowedDAOs
+}
+
+type Proposals struct {
+	seq       int64
+	*avl.Tree // *avl.Tree[ProposalID]*Proposal
+}
+
+func NewProposals() *Proposals {
+	return &Proposals{Tree: avl.NewTree()}
+}
+
+func (ps *Proposals) SetProposal(p *Proposal) ProposalID {
+	pid := ProposalID(ps.seq)
+	updated := ps.Set(pid.String(), p)
+	if updated {
+		panic("fatal error: Override proposals is not allowed")
+	}
+	ps.seq = ps.seq + 1
+	return pid
+}
+
+func (ps *Proposals) GetProposal(pid ProposalID) *Proposal {
+	pv, ok := ps.Get(pid.String())
+	if !ok {
+		return nil
+	}
+
+	return pv.(*Proposal)
+}
+
+// Author is an interface returned by the current govDAO implementation
+// representing the creator of a proposal. It is an interface
+// because we don't really know who/what will be allowed to create proposals
+// in the future, they can be any entity.
+type Author interface {
+	ID() string
+	String() string
+}
+
+type executor struct {
+	metadata   Metadata
+	callbackFn func(m Metadata) error
+}
+
+func newExecutor(fn func(m Metadata) error, metadata Metadata) *executor {
+	return &executor{
+		metadata:   metadata,
+		callbackFn: fn,
+	}
+}
+
+func (e *executor) String() string {
+	if e.metadata == nil {
+		return "No metadata available."
+	}
+
+	return e.metadata.String()
+}
+
+func (e *executor) Execute() error {
+	return e.callbackFn(e.metadata)
+}
+
+// Metadata contains any kind of information that will be stored on the executor
+// and will be sent to the callback function when the proposal is validated.
+// It can contains any kind of data.
+type Metadata interface {
+	IsMetadata()
```

Comment:
```suggestion
	AssertMetadata()
```

Reply:
I don't get the benefit of enforcing this interface; cna't it be an implementation detail of some DAOs, while what we expect isn't `func(Metadata) error` but just `func() error`?